Shopify/shopify-cli
# Retrieve latest CLI version in the background

### WHY are these changes introduced?

Inspired by #2052 



Currently we check for a new version as a blocking process. This isn't a good experience for CLI users, for whom the command hangs once daily until we check for a new version.

### WHAT is this pull request doing?



This approach will fork the process and check in the background, saving the result and warning the user the next time they run a command.

Note this isn't ideal for CLI environments which might run 1 command per CLI run and never show the warning. We might choose to mitigate by checking inline on non-TTY terminals.

### How to test your changes?



1. Enable [these lines](https://github.com/Shopify/shopify-cli/blob/cb0df573adb1aa9c07347d9401dfd2d974e12a60/lib/shopify_cli/core/entry_point.rb#L13-L14) in your dev environment (need to play with the `if/elsif` or just copy the lines to after the conditional)
2. Change the [version](https://github.com/Shopify/shopify-cli/blob/cb0df573adb1aa9c07347d9401dfd2d974e12a60/lib/shopify_cli/version.rb#L2) to something older
3. Change the [checking interval](https://github.com/Shopify/shopify-cli/blob/cb0df573adb1aa9c07347d9401dfd2d974e12a60/lib/shopify_cli/context.rb#L19) to 1 second so it'll check each time you run the command
4. Run `bin/shopify version` a few times and you'll start to see the new version warning (hopefully on round 2)

<img width="1441" alt="Screen Shot 2022-03-29 at 13 04 50" src="https://user-images.githubusercontent.com/6288426/160587482-f66610ea-b643-4761-971e-ae0ef3513b33.png">

### Update checklist
